### PR TITLE
Add cutouts for casda

### PR DESCRIPTION
**Pull request by jd-au**
Added the ability to produce spatial and spectral cutouts from ASKAP images and cubes using CASDA's SODA implementation.

### CODE REVIEW

**Review by bsipocz — COMMENTED** [review]
I mostly have nitpick comments. 

The bigger point is that I feel the module would benefit if the baseclass would be updated to be `QueryWithLogin` to help with admining the authentication, following the API of the other modules that deal with the need of authentication.

**Comment by jd-au** [comment]
> I mostly have nitpick comments.
> 
> The bigger point is that I feel the module would benefit if the baseclass would be updated to be `QueryWithLogin` to help with admining the authentication, following the API of the other modules that deal with the need of authentication.

It might be best to do that as a separate PR, however I will start on it immediately so that both changes are in the same release

**Comment by bsipocz** [comment]
>>    The bigger point is that I feel the module would benefit if the baseclass would be updated to be QueryWithLogin to help with admining the authentication, following the API of the other modules that deal with the need of authentication.

> It might be best to do that as a separate PR, however I will start on it immediately so that both changes are in the same release

Yes, I agree that it's beyond the scope here and would better be in a separate PR.

File: astroquery/casda/core.py [hunk 1]
```diff
@@ -102,23 +102,57 @@ def _args_to_payload(self, **kwargs):
...
+            elif kwargs.get('width') is not None and kwargs.get('height') is not None:
+                width = u.Quantity(kwargs['width']).to(u.deg).value
+                height = u.Quantity(kwargs['height']).to(u.deg).value
+                top = fk5_coords.dec.degree + (height/2)
+                bottom = fk5_coords.dec.degree - (height/2)
+                left = fk5_coords.ra.degree - (width/2)
+                right = fk5_coords.ra.degree + (width/2)
+                pos = 'RANGE {} {} {} {}'.format(left, right, bottom, top)
+            else:
+                pos = 'CIRCLE {} {} {}'.format(fk5_coords.ra.degree, fk5_coords.dec.degree, 1*u.arcmin.to(u.deg))
```
[the remark this answers is not in the document]
Comment (jd-au) [reply astroquery/casda/core.py]: I've moved this to the signature of the `query_region_async` , `_args_to_payload` and `cutout` functions.

File: astroquery/casda/core.py [hunk 2]
```diff
@@ -102,23 +102,57 @@ def _args_to_payload(self, **kwargs):
...
+                raise ValueError("Either 'channel' or 'band' values may be provided but not both.")
+
+            band = kwargs.get('band')
+            if not isinstance(band, (list, tuple)) or len(band) != 2:
+                raise ValueError("The 'band' value must be a list of 2 wavelength or frequency values.")
+            if (band[0] is not None and not isinstance(band[0], u.Quantity)) or (band[1] is not None and not isinstance(band[1], u.Quantity)):
+                raise ValueError("The 'band' value must be a list of 2 wavelength or frequency values.")
+            if band[0] is not None and band[1] is not None and band[0].unit.physical_type != band[1].unit.physical_type:
+                raise ValueError("The 'band' values must have the same kind of units.")
+            if band[0] is not None or band[1] is not None:
```
[the remark this answers is not in the document]
Comment (jd-au) [reply astroquery/casda/core.py]: I tried out a few approaches to simplifying these but none made it much better. I've pull out the `band[0] is not None` code into variables to hopefully make it more readable though.

File: astroquery/casda/core.py [hunk 3]
```diff
@@ -102,23 +102,57 @@ def _args_to_payload(self, **kwargs):
...
+                    # Swap the order when changing frequency to wavelength
+                    min_band = '-Inf' if band[1] is None else str(band[1].to(u.m, equivalencies=u.spectral()).value)
+                    max_band = '+Inf' if band[0] is None else str(band[0].to(u.m, equivalencies=u.spectral()).value)
+                else:
+                    raise ValueError("The 'band' values must be wavelengths or frequencies.")
+
+                request_payload['BAND'] = '{} {}'.format(min_band, max_band)
+
+        if kwargs.get('channel') is not None:
+            channel = kwargs.get('channel')
```
[the remark this answers is not in the document]
Comment (jd-au) [reply astroquery/casda/core.py]: Change made

File: astroquery/casda/core.py [hunk 4]
```diff
@@ -102,23 +102,57 @@ def _args_to_payload(self, **kwargs):
...
+            if not isinstance(channel, (list, tuple)) or len(channel) != 2:
+                raise ValueError("The 'channel' value must be a list of 2 integer values.")
+            if (not isinstance(channel[0], int)) or (not isinstance(channel[1], int)):
+                raise ValueError("The 'channel' value must be a list of 2 integer values.")
```
[the remark this answers is not in the document]
Comment (jd-au) [reply astroquery/casda/core.py]: Change made

File: astroquery/casda/core.py [hunk 5]
```diff
@@ -102,23 +102,57 @@ def _args_to_payload(self, **kwargs):
...
-            radius = u.Quantity(kwargs['radius']).to(u.deg)
-            pos = 'CIRCLE {} {} {}'.format(fk5_coords.ra.degree, fk5_coords.dec.degree, radius.value)
-        elif kwargs['width'] is not None and kwargs['height'] is not None:
-            width = u.Quantity(kwargs['width']).to(u.deg).value
-            height = u.Quantity(kwargs['height']).to(u.deg).value
-            top = fk5_coords.dec.degree - (height/2)
-            bottom = fk5_coords.dec.degree + (height/2)
-            left = fk5_coords.ra.degree - (width/2)
-            right = fk5_coords.ra.degree + (width/2)
-            pos = 'RANGE {} {} {} {}'.format(left, right, top, bottom)
```
[the remark this answers is not in the document]
Comment (jd-au) [reply astroquery/casda/core.py:117]: Change made - that's the first time I have encountered f-strings so thank-you for the suggestion.

File: astroquery/casda/core.py [hunk 6]
```diff
@@ -324,6 +419,26 @@ def _create_soda_job(self, authenticated_id_tokens, soda_url=None):
...
+        verbose: bool
+            Should progress be logged periodically
+        cutout_spec: map
+            The map containing the POS parameter defining the cutout.
+        """
+        if verbose:
+            log.info("Adding parameters: " + str(cutout_spec))
+        resp = self._request('POST', job_location + '/parameters', data=cutout_spec, cache=False)
+        resp.raise_for_status()
+        return
```
[the remark this answers is not in the document]
Comment (jd-au) [reply astroquery/casda/core.py]: Removed

File: astroquery/casda/core.py [hunk 7]
```diff
@@ -222,6 +237,86 @@ def stage_data(self, table, verbose=False):
...
+        coordinates : str or `astropy.coordinates`, optional
+            coordinates around which to produce a cutout, the radius will be 1 arcmin if no radius, height or width is provided.
+        radius : str or `astropy.units.Quantity`, optional
+            the radius of the cutout
+        height : str or `astropy.units.Quantity`, optional
+            the height for a box cutout
+        width : str or `astropy.units.Quantity`, optional
+            the width for a box cutout
+        band : list of `astropy.units.Quantity` with two elements, optional
+            the spectral range to be included, may be low and high wavelengths in metres or low and high frequencies in Hertz. Use None for an open bound.
```
[the remark this answers is not in the document]
Comment (jd-au) [reply astroquery/casda/core.py]: I've updated all code in this file to be 120 characters or less

File: docs/casda/casda.rst [hunk 8]
```diff
@@ -112,6 +112,40 @@ taken in scheduling block 2338 is shown below:
...
 
 
+Cutouts
+=======
+
+As well as accessing full data products, the CASDA service can produce cutout images and cubes from larger data products.
+The cutout support in AstroQuery allows both spatial and spectral cutouts.
+To produce a spatial cutout, pass in a coordinate and either a radius or a height and a width to the :meth:`~astroquery.casda.CasdaClass.cutout` method.
+To produce a spectral cutout, pass in either a band or a channel value. 
+For band, the value must be a list or tuple of two `astropy.units.Quantity` objects specifying a low and high frequency or wavelength. For an open ended range use `None` as the open value.
```
[the remark this answers is not in the document]
Comment (jd-au) [reply docs/casda/casda.rst]: I've updated the newly added content to be 120 characters or less

File: astroquery/casda/core.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bsipocz) [reply astroquery/casda/core.py:117]: yes, we couldn't really use them for a long time, but fortunately, we now dropped the old python versions that don't have it.

File: astroquery/casda/core.py [hunk 9]
```diff
@@ -222,6 +237,87 @@ def stage_data(self, table, verbose=False):
...
+            raise ValueError("Credentials must be supplied to download CASDA image data")
+
+        if table is None or len(table) == 0:
+            return []
+
+        job_url = self._create_job(table, 'async_service', verbose)
+
+        return self._complete_job(job_url, verbose)
+
+    def cutout(self, table, radius=1*u.arcmin, verbose=False, **kwargs):
```
Comment (bsipocz) [inline astroquery/casda/core.py:267]: I'm sorry, I think my previous comment was confusing. I meant that kwargs to be kwarg only (enforced by the `*` in the signature, not that they should be hidden in a `**kwargs`. 
```suggestion
    (self, table, *, coordinates=None, radius=None, height=None, width=None, band=None, channel=None, verbose=False)
```

**Review by bsipocz — APPROVED** [review]
Thank you! I'll commit the one misunderstanding in the previous review and merge once tests pass.

**Comment by jd-au** [comment]
> Thank you! I'll commit the one misunderstanding in the previous review and merge once tests pass.

Thanks for the review! If you have an example of the approach you want with kwarg I can make the change

**Comment by bsipocz** [comment]
Thanks @jd-au!

**Pull request merged**